geopandas/sindex.py

```diff
@@ -1,25 +1,341 @@
+                ]
+
+            elif predicate is not None:
+                tree_query = [
+                    i
+                    for i in tree_query
+                    if getattr(geometry, predicate)(self._geometries[i])
+                ]
+
+            if not sort:
+                return np.array(tree_query)  # unsorted
+
+            # sort
+            return np.sort(np.array(tree_query))
+
+        def query_bulk(self, geometry, predicate=None, sort=False):
+            """Compatibility layer for pygeos.query_bulk.
+
+            Iterates over `geometry` and queries index.
+            This operation is not vectorized and may be slow.
+            Use PyGEOS with `query_bulk` for speed.
+
+            Parameters
+            ----------
+            geometry : {GeoSeries, GeometryArray, numpy.array of PyGEOS geometries}
+                Accepts GeoPandas geometry iterables (GeoSeries, GeometryArray)
+                or a numpy array of PyGEOS geometries.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                The predicate to use for testing geometries from the tree
+                that are within the input geometry's envelope.
+                See `pygeos.strtree` documentation for more details on
+                expected behavior of predicates.
+            sort : bool, default False
+                Determines if the results should be sorted lexicographically using
+                geometry's indexes as the primary key and the sindex's indexes as the
+                secondary key.
+
+            Returns
+            -------
+            ndarray with shape (2, n)
+                The first subarray contains input geometry indexes.
+                The second subarray contains tree geometry indexes.
+            """  # noqa: E501
+            # Iterates over geometry, applying func.
+            tree_index = []
+            geo_index = []
+
+            for i, geo in enumerate(geometry):
+                res = self.query(geo, predicate=predicate, sort=sort)
+                if res.size > 0:
+                    # sort results and append
+                    tree_index.extend(res)
+                    geo_index.extend([i] * len(res))
+            return np.vstack([geo_index, tree_index])
+
+        @property
+        def size(self):
+            return len(self.leaves()[0][1])
+
+        @property
+        def is_empty(self):
+            return self.size == 0
+
+
+if compat.HAS_PYGEOS:
+
+    from pygeos import STRtree, box, points, Geometry  # noqa
+
+    class PyGEOSSTRTreeIndex(STRtree):
+        """
+        A simple wrapper around pygeos's STRTree
+        """
+
+        with_objects = namedtuple("with_objects", "object id")
+
+        def __init__(self, geometry):
+            # for compatibility with old RTree implementation, store ids/indexes
+            original_indexes = geometry.index
```

Comment:
This only works if `geometry` is a `GeoSeries` right?  `index` is not defined for `GeometryArray` / `ndarray`.

Reply:
Right. There was some discussion with @jorisvandenbossche as to what these wrappers should accept as input. Previously, it was a stream of bounds and indexes. I decided to make it a GeoSeries now. It could be expanded to ndarray or others, but things could get complicated, especially if we want to support operating on label indexes instead of iloc style indexes.

Reply:
I don't think we need to worry about the pandas index here, since the STRtree works with positional indices. So I would drop this for now.

We do need to be a bit careful about the the passed `geometry` though, as also the next line will only work if `geometry` is a GeoSeries (which is probably what we pass in internally?) 
A small helper function that ensures that any input (GeoSeries, GeometryArray, ndarray of geometries) gets converted to a GeometryArray might be useful to add for such use cases.

Reply:
The reason we need the original indexes is because the existing `rtree` wrapper _does_ have the option to work with the `loc/label` based indexes by storing them in the tree itself and retrieving them using the `objects` parameter ([source in master](https://github.com/geopandas/geopandas/blob/878a770ea9b976b693b982503f5af33d50b070c0/geopandas/base.py#L109)). If we want full compatibility with the existing wrapper, we need to store the pandas index somehow so that we can emulate the `objects` parameter. It would be easier if we could just drop support for that for now, which is what I was asking in my previous comment ([link](https://github.com/geopandas/geopandas/pull/1343#discussion_r410988037)).

I'd be happy to add a helper function like you describe if you think that should be within the scope of this PR.